caffe2/core/predictor.cc

```diff
@@ -48,9 +48,11 @@ void Predictor::run(const TensorVector& inputs, TensorVector* outputs) {
 
   CAFFE_ENFORCE(ws_.RunNet(run_net_.name()));
 
-  outputs->resize(run_net_.external_output_size());
-  for (auto i = 0; i < outputs->size(); ++i) {
-    (*outputs)[i] = extractOutputTensor(&ws_, run_net_.external_output(i));
+  if (outputs != nullptr) {
```

Comment:
This predictor is not very smart actually, as you can see it doesn't do much logic on top of normal Caffe2 Net interfaces. Which are perfectly fine to use in your c++ code directly. 

So if it doesn't give you too  much if would indeed have the Predictor code as simple as possible. Idea of predictor is that you don't need to access anything except inputs and outputs. If you do - you better to work with a Net interfaces IMO.

Its original purpose was to serve these two use-case:

1. Mobile(lightweight simple Predictor)
2. Python bindings

We plan to open-source another predictor which does smarter logic with threading and initialization nets and uses predictor_exporter format single file format (see corresponding file in the repo)